internal/core/core.go

```diff
@@ -506,7 +509,34 @@ func (cs *coreService) SendMultipartMessage(
 		return 0, errors.New("You must provide the ID of the user sending the message")
 	}
 
-	requestBody, err := common.CreateRequestBody(map[string]interface{}{"parts": options.Parts})
+	requestParts := make([]interface{}, len(options.Parts))
+
+	errChan := make(chan error)
+	uploads := 0
+
+	for i, part := range options.Parts {
+		switch p := part.(type) {
+		case NewAttachmentPart:
+			uploads++
+			go func(i int, p NewAttachmentPart) {
+				uploadedPart, err := cs.uploadAttachment(ctx, options.SenderID, options.RoomID, p)
+				requestParts[i] = uploadedPart
+				errChan <- err
+			}(i, p)
+		default:
+			requestParts[i] = part
+		}
+	}
+
+	for ; uploads > 0; uploads-- {
+		if err := <-errChan; err != nil {
+			return 0, fmt.Errorf("Failed to upload attachment: %v", err)
+		}
+	}
```

Comment:
I think you want an `errgroup` https://godoc.org/golang.org/x/sync/errgroup

Reply:
ah yes. That is exactly what I want.